ui/app/components/keymgmt/provider-edit.js

```diff
@@ -0,0 +1,108 @@
+import Component from '@glimmer/component';
+import { inject as service } from '@ember/service';
+import { action } from '@ember/object';
+import { tracked } from '@glimmer/tracking';
+import { task } from 'ember-concurrency';
+import { waitFor } from '@ember/test-waiters';
+
+/**
+ * @module KeymgmtKeyEdit
+ * ProviderKeyEdit components are used to display KeyMgmt Secrets engine UI for Key items
+ *
+ * @example
+ * ```js
+ * <KeymgmtProviderEdit @model={model} @mode="show" />
+ * ```
+ * @param {object} model - model is the data from the store
+ * @param {string} mode - mode controls which view is shown on the component - show | create |
+ * @param {string} [tab] - Options are "details" or "keys" for the show mode only
+ */
+
+export default class KeymgmtKeyEdit extends Component {
+  @service router;
+  @service flashMessages;
+
+  constructor() {
+    super(...arguments);
+    // key count displayed in details tab and keys are listed in keys tab
+    if (this.args.mode === 'show') {
+      this.fetchKeys.perform();
+    }
+  }
+
+  @tracked modelValidations;
+
+  get isShowing() {
+    return this.args.mode === 'show';
+  }
+  get isCreating() {
+    return this.args.mode === 'create';
+  }
+  get viewingKeys() {
+    return this.args.tab === 'keys';
+  }
+
+  @task
+  @waitFor
+  *saveTask() {
+    const { model } = this.args;
+    try {
+      yield model.save();
+      this.router.transitionTo('vault.cluster.secrets.backend.show', model.id, {
+        queryParams: { itemType: 'provider' },
+      });
+    } catch (error) {
+      this.flashMessages.danger(error.errors.join('. '));
+    }
+  }
+  @task
+  @waitFor
+  *fetchKeys(page = 1) {
+    try {
+      yield this.args.model.fetchKeys(page);
+    } catch (error) {
+      this.flashMessages.danger(error.errors.join('. '));
+    }
+  }
+
+  @action
+  async onSave(event) {
+    event.preventDefault();
+    const { validations } = await this.args.model.validate();
+    if (validations.isValid) {
+      this.saveTask.perform();
+    } else {
+      // FormField expects validationMessages in a shape not output by ember-cp-validations
+      this.modelValidations = validations.errors.reduce(
```

Comment:
Since we use FormField all over the place, I wonder if this is worth putting into a helper for use in other areas later?

Reply:
Yes I think that's a good idea for the time being. I will move it out to a util that can be imported. It might be best in the future to look at refactoring the `FormField` component to accept this shape for validation errors if they are always coming from models.

Reply:
[I created a helper and refactored it a bit](https://github.com/hashicorp/vault/pull/13797/commits/f9c090d72d3f578894ab42a83c7c011fdcb57ea9) since I had specific logic to build the object based on the nested credentials object. Now it will build the error object for any level of nested attributes.